### PR TITLE
Expose targets and versions for more dependencies

### PR DESCRIPTION
**Pull request by DomClark**
The main changes are to the following find modules:
* `FindLame.cmake`: expose version
* `FindOggVorbis.cmake`: convert to use imported targets, expose version
* `FindPortAudio.cmake`: expose version
* `FindSamplerate.cmake`: convert to use imported targets, expose version
* `FindSndFile.cmake`: convert to use imported targets, expose version

I have added a helper function `_find_package_config_mode_with_fallback`, since the same general pattern was used in several find modules. I have also updated `FindSTK.cmake`, which followed that pattern, to use the new function, although nothing new has been exposed.

Versions are taken from config mode or pkg-config where available, but I have provided a fallback for when neither of those are available, or they are available but don't specify the version. The libraries relevant to this PR don't have a version string available in the header, instead requiring the user to call a function in the library itself to get the version (apart from STK, which doesn't seem to expose its version at all). I have added another helper function `_determine_version_from_source` to aid in this scenario.

Direct references to the include directories or libraries of the affected dependencies have been removed, and instead replaced with links to the imported targets, as is modern CMake style. This necessitated a few further changes as some plugins no longer implicitly had access to header files used by the core. All plugins now link to the core `lmms` executable target, and as such any interface link libraries of that target are correctly propagated. This was already the case for Windows builds, but now applies on macOS and Linux too. This allowed getting rid of some custom linker flags for those platforms, which are now handle automatically by CMake: on macOS, `-bundle_loader` is no longer explicitly passed when linking plugins, and on Linux, `-Wl,-E` is no longer explicitly passed when building the core. Instead, the plugin-core dependency and `ENABLE_EXPORTS` property cause these to be added.

This PR was primarily motivated by the need to get the version of libsndfile for #6750, but also works towards more modern CMake usage in general. The particular dependencies touched here are those that are in use on Windows and already have find modules in place, with the exception of FFTW which was more complex. I plan to cover more in the future.

### CODE REVIEW

**Comment by tresf** [comment]
I'm not sure how scalable this is, but LAME's version is available using:

```bash
ar p libmp3lame.a version.o
# binary junk ... LAME3.100 ...binary junk
```

It's a binary output (tested on Mac and Linux).  Linux can actually use `tr` to grab the version:

```bash
ar p /usr/lib/aarch64-linux-gnu/libmp3lame.a version.o |tr '\0' '\n' |grep -a LAME
# LAME3.100 
```

```bash
ar p /usr/lib/aarch64-linux-gnu/libvorbis.a info.o |tr '\0' '\n' |grep -a libVorbis
# Xiph.Org libVorbis I 20200704 (Reducing Environment)
# Xiph.Org libVorbis 1.3.7
```

The reason I mention this is because it may provide a compile-less option for CMake to get version information.  Of course, the `tr '\0' '\n'` and `grep` would have to be reproduced CMake-side.

**Comment by DomClark** [comment]
It does avoid problems with running compiled code when cross-compiling, which is nice, but I don't feel it's as robust - there's no guarantee that the string will remain in a single piece in the binary, neatly surrounded with null characters. Using an official interface to the library should keep us safe within the same major version. (Admittedly, we're still relying on the specific format of the information returned by that interface, but the binary solution has that same problem too.)

It's not a general solution either, as it won't work for libraries that only expose numeric version information. I'm not sure whether that applies to any of the libraries in this PR, but it's a concern for similar future changes, which I intend to make.

(Not really an argument, but it also just feels really dirty.)

**Comment by tresf** [comment]
> (Not really an argument, but it also just feels really dirty.)

I agree and I don't want to beat up the topic too much, but... 

<details>
<summary>
Click to expand ...
</summary>

> It's not a general solution either, as it won't work for libraries that only expose numeric version information.

Do you know which ones? PortAudio offers a string format as well, just needs to be parsed.

```
ar p /usr/lib/aarch64-linux-gnu/libportaudio.a pa_front.o | tr '\0' '\n' |grep PortAudio
# PortAudio V19.6.0-devel, revision 396fe4b6699ae929d3a685b3ef8a7e97396139a4
# PortAudio not initialized
# Internal PortAudio error
```

One sign that it may not work correctly (could be an AR compatibility though), when I try to look inside `libsamplerate.dll.a`, from a Linux/Unix platform, I get the following:

```
ar t /Users/owner/Downloads/libsamplerate-mingw-w64_0.1.9-3_all/usr/x86_64-w64-mingw32/lib/libsamplerate.dll.a 
# d000031.o/
# d000002.o/
# ...
```

... `DUMPBIN.EXE` / `LIB.EXE` not being much better... 
```
File Type: LIBRARY

  Summary

          14 .idata$2
          E8 .idata$4
```

... so I'm not sure how well this would work for cross-compilation (I didn't test with any tools like mingw, but they generally have the pkgconfig information already available anyway, so I'm not sure how useful it would really be.
</details>

File: cmake/modules/BuildPlugin.cmake
Comment (sakertooth) [inline cmake/modules/BuildPlugin.cmake:59]: Kind of curious why the original code was only linking ``lmms`` for Windows builds (I'm guessing the plugins needed to see certain symbols on Windows)?

File: cmake/modules/ImportedTargetHelpers.cmake
Comment (sakertooth) [inline cmake/modules/ImportedTargetHelpers.cmake:115]: I think we could use a status message here, explaining why an early return was made.

**Review by sakertooth — APPROVED** [review]
LGTM, just a few thoughts/questions.

File: cmake/modules/BuildPlugin.cmake
Comment (DomClark) [reply cmake/modules/BuildPlugin.cmake:59]: That's right. If you reference an undefined symbol on Linux, the linker assumes it'll be available at runtime. This doesn't work on Windows - there's no global symbol table, so the linker needs to know which library each symbol will come from.

Comment (tresf) [reply cmake/modules/BuildPlugin.cmake:59]: I'm interested in this decision too.  I don't know enough about the symbol tables, but I'm curious what advantages (disadvantages) to toggle this for all systems.

I tried to dig around CMake docs to understand this better and just got more confused.  I did stumble upon this while I was searching though, which I found to be very interesting, perhaps unrelated to this PR https://stackoverflow.com/q/59838565/3196753.

**Comment by tresf** [comment]
This is more of a style guideline feedback, but this is the first time we've introduced underscores for shared functions in CMake.  I'm sure this is copying or honoring a convention, but I wanted a place to discuss this, since other project maintainers will want to follow suit.

My **assumption** is that the underscore dictates that these functions are specifically reserved from within `Find<Foo>.cmake` scripts and never parent scripts, however this PR is the first instance I see of this (I don't work with larger CMake projects like `vcpkg`, so if it's very common elsewhere, I may be blind to this).

My only intention here is to discuss and establish the standard moving forward, I have no helpful feedback to provide on a better approach unless we use namespacing. ❤️

File: cmake/modules/ImportedTargetHelpers.cmake
Comment (tresf) [inline cmake/modules/ImportedTargetHelpers.cmake:86]: I would recommend renaming this function to mimic the tool's proper name `pkg-config`.  I think this is a bit easier to read, given the abundance of the word "package" in helper scripts. 

```diff
- _find_package_config_mode_with_fallback
+ _find_pkg_config_mode_with_fallback
```

The variables may also benefit form a rename, but I couldn't think of a suggestion that was better than what is already provided.

Comment (tresf) [reply cmake/modules/ImportedTargetHelpers.cmake:86]: > `# The package to search for with config mode`

Ok, this can be disregarded, I realize that the function was named after [CMake's config mode](https://cmake.org/cmake/help/latest/command/find_package.html#config-mode-search-procedure) first, and `pkg-config` as the fallback.  Safe to resolve.

**Comment by tresf** [comment]
> Direct references to the include directories or libraries of the affected dependencies have been removed, and instead replaced with links to the imported targets

So does this force all plugins to implicitly include the headers from `lmms`?  This may not be the best place to ask rudimentary "modern CMake style" questions, but what's particularly interesting is these changes, especially how many samplerate changes were made as a result of this change.

**Comment by tresf** [comment]
```cmake
_find_package_config_mode_with_fallback(SampleRate SampleRate::samplerate
	LIBRARY_NAMES "samplerate" "libsamplerate" "libsamplerate-0"
	INCLUDE_NAMES "samplerate.h"
	PKG_CONFIG samplerate
	PREFIX Samplerate
)
```

Just typo-hunting.  Can you help explain the reason for `SampleRate::`, specifically the capital "R" here?  I noticed some libraries are found with lowercase, camelcase etc.  I'm asking as a general question (instead of in-line) because I think understanding this will help others use the helpers. 🍻

**Comment by DomClark** [comment]
> This is more of a style guideline feedback, but this is the first time we've introduced underscores for shared functions in CMake.  I'm sure this is copying or honoring a convention, but I wanted a place to discuss this, since other project maintainers will want to follow suit.
>
> My **assumption** is that the underscore dictates that these functions are specifically reserved from within `Find<Foo>.cmake` scripts and never parent scripts, however this PR is the first instance I see of this (I don't work with larger CMake projects like `vcpkg`, so if it's very common elsewhere, I may be blind to this).

That's right - see https://stackoverflow.com/a/60626774, for example. 

> So does this force all plugins to implicitly include the headers from lmms?

Specifically, it adds the include paths for the headers of LMMS and its dependencies to compiler invocations for all plugins. The plugins still need to `#include` the headers they use. CMake will look at a target's transitive dependencies (i.e. its dependencies, the dependencies of those dependencies, and so on), and combine their include directories to compute the include directories to use for that target. It does a similar thing for macro defintions and link libraries.

This is the essence of modern CMake. Include directories, link libraries, compile definitions, etc. are associated with targets, and dependencies are specified between targets. Then the appropriate directories/libraries/definitions are used automatically, without needing to manually add them in every place they're required.

> But what's particularly interesting is these changes, especially how many samplerate include changes were made as a result of this change.

Yeah - libsamplerate is something we use all over the place, so there are a lot of changes associated with it. Most dependencies are only used in the core, or only used in a single plugin, so don't have many corresponding changes in the list files. Outside of the find module, the samplerate changes are all quite similar - removing manual specification of link libraries and include directories, and replacing them with a dependency on the samplerate target.

>  Can you help explain the reason for `SampleRate::`, specifically the capital "R" here? I noticed some libraries are found with lowercase, camelcase etc. I'm asking as a general question (instead of in-line) because I think understanding this will help others use the helpers.

It's the name used by libsamplerate itself - see https://github.com/libsndfile/libsamplerate/blob/22bd06eb114850ebe31981eb794d150a95439fef/src/CMakeLists.txt#L166. Deciding what target name to use is a matter of looking at the library's code. For anyone using vcpkg, however, it usually prints the appropriate target name for you after installation, e.g.
```
> vcpkg install libsamplerate
...
libsamplerate provides CMake targets:

    # this is heuristically generated, and may not be correct
    find_package(SampleRate CONFIG REQUIRED)
    target_link_libraries(main PRIVATE SampleRate::samplerate)
```

File: cmake/modules/ImportedTargetHelpers.cmake
Comment (DomClark) [reply cmake/modules/ImportedTargetHelpers.cmake:115]: I'd like this function to stay quiet for now - you'll notice it passes `QUIET` to `find_package` and `pkg_check_modules`. The find module that calls it should first search for all the dependencies, and if it fails to find them, that's where the error should be reported. Perhaps it could be useful to have a `QUIET` argument for this function too, and report any errors if it's not passed, but I'd rather do that if or when we need it.

File: cmake/modules/BuildPlugin.cmake
Comment (DomClark) [reply cmake/modules/BuildPlugin.cmake:59]: The immediate advantage is that dependencies become easier to manage, as any public dependencies of the core will be picked up automatically by plugins. It also allows plugins to consume symbols from the core in a consistent way across platforms (previously we only used this approach for Windows, and used some custom flags for macOS - the core dependency doesn't need to be specified on Linux as it only looks up the symbols at run time). I'm not aware of any disadvantages, and everything still seems to work, but that of course doesn't mean there aren't any.

Comment (tresf) [reply cmake/modules/BuildPlugin.cmake:59]: My first thought is that they stop being "plugins" from the sense that they're dependant on another library now.  Is this a misconception?

Comment (DomClark) [reply cmake/modules/BuildPlugin.cmake:59]: It's not wrong, but I don't think it's relevant. The dependency was always present, just not as an explicit target-target dependency in CMake (except for Windows builds). IIRC, there have been discussions elsewhere about making the plugin API purer, so it doesn't require linking to the core, but that's not the goal here.

Comment (tresf) [reply cmake/modules/BuildPlugin.cmake:59]: > The dependency was always present, just not as an explicit target-target dependency in CMake (except for Windows builds)

I think of projects like https://github.com/pgiblock/lmms-lv2 are an interesting step in that direction.  If as you state, it's already a transient dependency, it's certainly off-topic for this discussion.

**Comment by sakertooth** [comment]
@tresf do you approve of this? This PR looks ready to merge.

**Comment by tresf** [comment]
> > My assumption is that the underscore dictates that these functions are specifically reserved from within Find<Foo>.cmake scripts and never parent scripts, however this PR is the first instance I see of this


> That's right - see https://stackoverflow.com/a/60626774, for example

Hmm... I typed a bunch of stuff about underscores and decided to delete it, TL;DR.  The short is, I don't agree that they're helpful when exposed to FindXXX.cmake calls, and personally have no intention to prefix shared functions this way moving forward.  I'd love a second opinion.

That said, it's not worth holding up the merge over. 🍻

**Review by tresf — APPROVED** [review]

**Pull request merged**
